File: SocialChallenge/Categories/UINavigationBar/UINavigationBar+JSCCustomHeight.m
//
//  UINavigationBar+JSCCustomHeight.m
//  SocialChallenge
//
//  Created by Javier Laguna on 28/02/2016.
//
//

#import "UINavigationBar+JSCCustomHeight.h"

#import "objc/runtime.h"

/**
 constant for the height.
 */
static char *const kJSCHeightKey = "Height";

@implementation UINavigationBar (JSCCustomHeight)

- (void)jsc_setHeight:(CGFloat)height
{
    objc_setAssociatedObject(self,
                             kJSCHeightKey,
                             @(height),
                             OBJC_ASSOCIATION_RETAIN_NONATOMIC);
}

#pragma mark - Getters

- (NSNumber *)height
{
    return objc_getAssociatedObject(self,
                                    kJSCHeightKey);
}

#pragma mark - System

- (CGSize)sizeThatFits:(CGSize)size
{
    CGSize newSize;
    
    if (self.height)
    {
        newSize = CGSizeMake(self.superview.bounds.size.width,
                             [self.height floatValue]);
    }
    else
    {
        newSize = [super sizeThatFits:size];
    }
    
    return newSize;
}

@end

File: SocialChallenge/Categories/UINavigationBar/UINavigationBar+JSCCustomHeight.swift
//
//  UINavigationBar+JSCCustomHeight.swift
//  SocialChallenge
//
//  Created by Javier Laguna on 18/03/2016.
//
//

import Foundation
import UIKit

/**
constant for the height.
*/
var AssociatedHeightHandle: UInt8 = 0

/**
allows to define a custom size for the navigation bar.
*/
extension UINavigationBar {
    
    var height: CGFloat {
        
        get {
            
            if let unwrappedHeight = objc_getAssociatedObject(self, &AssociatedHeightHandle)
            {
                return unwrappedHeight as! CGFloat
            }
            else {
                
                return 0.0
            }
        }
        set {
            
            objc_setAssociatedObject(self, &AssociatedHeightHandle, newValue, objc_AssociationPolicy.OBJC_ASSOCIATION_RETAIN_NONATOMIC)
        }
    }

    /**
     Sets the height of the Navigation Bar.
     
     @param height the height.
     */
    func jsc_setHeight(height: CGFloat) {
        
        self.height = height
    }
    
    //MARK: System
    
    override public func sizeThatFits(size: CGSize) -> CGSize {
        
        var newSize: CGSize = super.sizeThatFits(size)
        
        if (height > 0)
        {
            newSize = CGSizeMake(newSize.width, height)
        }
        
        return newSize
    }
    
}